rllib/policy/sample_batch.py

```diff
@@ -133,27 +134,68 @@ def __init__(self, *args, **kwargs):
         if self._is_training is None:
             self._is_training = self.pop("is_training", False)
 
+        # Try to infer the "length" of the SampleBatch by finding the first
+        # value that is actually a ndarray/tensor.
         lengths = []
         copy_ = {k: v for k, v in self.items() if k != SampleBatch.SEQ_LENS}
         for k, v in copy_.items():
             assert isinstance(k, str), self
 
-            # TODO: Drop support for lists as values.
+            # TODO: Drop support for lists and Numbers as values.
             # Convert lists of int|float into numpy arrays make sure all data
             # has same length.
-            if isinstance(v, list):
+            if isinstance(v, (Number, list)):
                 self[k] = np.array(v)
 
-            # Try to infer the "length" of the SampleBatch by finding the first
-            # value that is actually a ndarray/tensor. This would fail if
-            # all values are nested dicts/tuples of more complex underlying
-            # structures.
+            if k == SampleBatch.INFOS:
+                # Don't attempt to count on infos since we make no assumptions
+                # about its content
+                continue
+
+            # If this is a nested dict (for example a nested observation),
+            # try to flatten it, assert that all elements have the same length (batch
+            # dimension)
+            v_list = tree.flatten(v) if isinstance(v, (dict, tuple)) else [v]
+            # TODO: Drop support for lists and Numbers as values.
+            # If v_list contains lists or Numbers, convert them to arrays, too.
+            v_list = [
+                np.array(_v) if isinstance(_v, (Number, list)) else _v for _v in v_list
+            ]
             try:
-                len_ = len(v) if not isinstance(v, (dict, tuple)) else None
+                # Add one of the elements' length, since they are all the same
+                len_ = len(v_list[0])
```

Comment:
use ``_len``? underscore suffix is not part of the style guide,

Reply:
Done!